src/network/networkd-link.c

```diff
@@ -3084,6 +3084,9 @@ static int link_carrier_gained(Link *link) {
 
         assert(link);
 
+        if (link->state == LINK_STATE_CONFIGURED && link->network->ignore_carrier_gainloss)
+                return 0;
+
```

Comment:
Could you please explain how do we get into this scenario with regard to line 3117 ?

Reply:
Say when cable is removed - thencarrier is lost and link_carrier_lost() is called. Now if IgnoreCarrier/RemainWithoutCarrier is true, then we dont do anything i.e we do not drop config.

Now when cable is connected - link_carrier_gained() is called. But since we didn't drop config - we do not need to reconfigure.

Reply:
yes right I guess you mentioned this is only for static configuration not for DHCP stuff ?

Reply:
It may be used for other purposes when administrator knows that config / address / routes etc will not change on link carrier changes. (like DHCP server always gives same IP based on MAC address).

if everything is fine then please give your approval to developers so that they can consider merging this PR.

Thank you.

Reply:
No for DHCP this won't work did you tested ?

Reply:
Its not really meant for DHCP anyway. I was just giving example on how it may be used for other purposes if administrator knows what he is doing.

Reply:
1. 5 sec window to DHCP lease expire.
2. lost carrier.
3. the conf is retained but dhcp can't renew the address since there is no carrier. so it's dropped the address etc when the lease expired. (for this we have 'dhcpcritical')
4. Now gained carrier and in the `gained_carrier` DHCP can't get the new address since in the gained carrier you asked to return. So only the static stuff will work.

For this reason let's make it specific to static only and  code should move down just above drop_config.

Again do you still think we need the code in the `*gained_carrier`

Reply:
If it is DHCP where IP is really dynamic - why would administrator set IgnoreCarrierGainLoss=true?

There is no reason to set it and hence the case above that you have mentioned will not occur.

So for me current code is correctly placed.

And yes code in gained_carrier is needed otherwise when the carrier is back - it will again acquire_config and again set address and do all sorts of things again (which is already done when link was up the first time)

May be @poettering can give his thoughts?

Reply:
Why wound not admin will not set to retain the static configuration. Please test your code does it even work for dhcp case

Reply:
Thats what I am saying - its not supposed to work in DHCP because its not really meant for DHCP. If its DHCP then there is no reason to set IgnoreCarrierGainLoss=.

Reply:
This implementation is breaking the other feature Please fix that.

Reply:
I have no idea what you mean by breaking the other feature, **which feature**?!

Here is my log of my testing: (I do not see anything failing with static as well as DHCP)

# Case 1) Static IP
```
[Network]
Address=192.168.1.31/24
Gateway=192.168.1.1
IgnoreCarrierGainLoss=true
ConfigureWithoutCarrier=true
```
**a) System boot: Static IP - Cable connected - eth0 Configured**
Dec 02 20:28:59 mysystem systemd[1]: Started Network Service.
Dec 02 20:28:59 mysystem systemd-networkd[438]: eth0: IPv6 successfully enabled
Dec 02 20:28:59 mysystem systemd-networkd[438]: eth0: Gained carrier
Dec 02 20:29:01 mysystem systemd-networkd[438]: eth0: Gained IPv6LL
Dec 02 20:29:13 mysystem systemd-networkd[438]: eth0: Configured

> $ ip addr list dev eth0
> 2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc htb state UP group default qlen 1000
>     link/ether 08:00:27:84:1d:41 brd ff:ff:ff:ff:ff:ff
>     inet 192.168.1.31/24 brd 192.168.1.255 scope global eth0
>        valid_lft forever preferred_lft forever

**b) Cable removed - Static IP - Lost carrier - but IP remains as expected**
Dec 02 20:59:25 mysystem kernel: e1000: eth0 NIC Link is Down
Dec 02 20:59:25 mysystem systemd-networkd[3382]: eth0: Lost carrier
Dec 02 20:59:25 mysystem systemd-timesyncd[218]: No network connectivity, watching for changes.

> $ ip addr list dev eth0
> 2: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc htb state DOWN group default qlen 1000
>     link/ether 08:00:27:84:1d:41 brd ff:ff:ff:ff:ff:ff
>    inet 192.168.1.31/24 brd 192.168.1.255 scope global eth0
>        valid_lft forever preferred_lft forever

**c) Cable re-connected - Static IP - Gained carrier - not reconfigured unnecessarily**
Dec 02 21:00:46 mysystem systemd-networkd[3382]: eth0: Gained carrier
Dec 02 21:00:46 mysystem kernel: e1000: eth0 NIC Link is Up 1000 Mbps Full Duplex, Flow Control: RX
Dec 02 21:00:46 mysystem systemd-timesyncd[218]: Network configuration changed, trying to establish connection.
Dec 02 21:01:44 mysystem systemd-timesyncd[218]: Synchronized to time server 123.108.200.124:123 (0.arch.pool.ntp.org).

> $ ip addr list dev eth0
> 2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc htb state UP group default qlen 1000
>     link/ether 08:00:27:84:1d:41 brd ff:ff:ff:ff:ff:ff
>     inet 192.168.1.31/24 brd 192.168.1.255 scope global eth0
>        valid_lft forever preferred_lft forever

# Case 2) DHCP

_**Note** that as I said earlier, IgnoreCarrierGainLoss makes no sense here, but I still set it to true for testing_

```
[Network]
DHCP=ipv4
IgnoreCarrierGainLoss=true
ConfigureWithoutCarrier=true
```

**a) System boot: DHCP - Cable connected - DHCP Configured**
Dec 02 21:25:20 mysystem systemd[1]: Started Network Service.
Dec 02 21:25:20 mysystem systemd-networkd[433]: eth0: IPv6 successfully enabled
Dec 02 21:25:20 mysystem systemd-networkd[433]: eth0: Gained carrier
Dec 02 21:25:21 mysystem systemd-networkd[433]: eth0: Gained IPv6LL
Dec 02 21:25:23 mysystem systemd-networkd[433]: eth0: DHCPv4 address 192.168.1.33/24 via 192.168.1.1

> $ ip addr list dev eth0
> 2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc htb state UP group > default qlen 1000
>     link/ether 08:00:27:84:1d:41 brd ff:ff:ff:ff:ff:ff
>     inet 192.168.1.33/24 brd 192.168.1.255 scope global dynamic eth0
>        valid_lft 86176sec preferred_lft 86176sec

**b) Cable removed - DHCP - Lost carrier - DHCP lease lost**
Dec 02 21:33:33 mysystem kernel: e1000: eth0 NIC Link is Down
Dec 02 21:33:33 mysystem systemd-networkd[433]: eth0: Lost carrier
Dec 02 21:33:33 mysystem systemd-networkd[433]: eth0: DHCP lease lost

> $ ip addr list dev eth0
> 2: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc htb state DOWN group default qlen 1000
>     link/ether 08:00:27:84:1d:41 brd ff:ff:ff:ff:ff:ff
>     inet6 fe80::a00:27ff:fe84:1d41/64 scope link
>        valid_lft forever preferred_lft forever

**c) Cable removed - DHCP - Gained carrier - DHCP IP returns**
Dec 02 21:36:43 mysystem kernel: e1000: eth0 NIC Link is Up 1000 Mbps Full Duplex, Flow Control: RX
Dec 02 21:36:43 mysystem systemd-networkd[433]: eth0: Gained carrier
Dec 02 21:36:46 mysystem systemd-networkd[433]: eth0: DHCPv4 address 192.168.1.33/24 via 192.168.1.1
Dec 02 21:36:46 mysystem systemd-networkd[433]: eth0: Configured
Dec 02 21:38:12 mysystem systemd-timesyncd[229]: Synchronized to time server 123.108.200.124:123 (2.arch.pool.ntp.org).

> $ ip addr list dev eth0
> 2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc htb state UP group default qlen 1000
>     link/ether 08:00:27:84:1d:41 brd ff:ff:ff:ff:ff:ff
>     inet 192.168.1.33/24 brd 192.168.1.255 scope global dynamic eth0
>        valid_lft 86314sec preferred_lft 86314sec

Reply:
Let me try again.

Case 1: has no issue for static config.

Case2: DHCP:
   In a ideal scenario when the carrier lost we drop the DHCP config and stop the client. On carrier gain we again start the client immediately  and we get the config from DHCP and we configure. There is no *delay*.

 With you patch we don't stop the client and the client keep on running. The client tries to send 
DISCOVER . If no reply we increase the TMO and we send DISCOVERER.  Assuming for sometime no carrier gain.   
```
DHCP CLIENT (0xae3b7174): REQUEST (renewing)
NDISC: Sent Router Solicitation, next solicitation in 1min 3s
DHCP CLIENT (0xae3b7174): REQUEST (rebinding)
DHCP CLIENT (0xae3b7174): EXPIRED
enp0s9: DHCP lease lost
Setting transient hostname: 'n/a'
```
Here is the period your code will fail to retain the DHCP config.


Here is the catch just before the lease expired you lost the carrier. But with your code client will be still running  and eventually it will drop the config because of TMO I just explained . Since after sometime you got carrier gain it won't start the DHCP client immediately. It will wait for the TMO to resend the DISCOVERER. 

For that period even if you have carrier you can't talk to DHCP because client is waiting for the TMO. In a network router scenario even if for 5 sec you can't afford to do that. 

So this patch can't stop DHCP client to to drop the config and whenever we gain a carrier it fails to act immediately to get the config from the DHCP server.

Reply:
But I fail to understand why would administrator set IgnoreCarrierGainLoss for **case 2**???

Its not supposed to be set to true. If its not set to true then above scenario will never occur.

Reply:
Well we just can't predict anything you know. We just have to careful that this should not happen.
If this is not for case 2 then why write code for that which affects it. 
Again case 2 is getting involved if you consider for case 1 . It's not someone purposefully doing this .

Reply:
I dont think its common to mix DHCP and static in config.

But even if its common to write this:

```
[Network]
DHCP=ipv4
Address=192.168.1.31/24
Gateway=192.168.1.1
```

No way common to add IgnoreCarrierGainLoss= in config. Nor it will become common for DHCP in future,

In the end, it is administrator's job to carefully set configuration and not put things randomly.